source/common/quic/http_datagram_handler.cc

```diff
@@ -0,0 +1,86 @@
+#include "source/common/quic/http_datagram_handler.h"
+
+#include "source/common/common/logger.h"
+#include "source/common/http/header_map_impl.h"
+
+#include "absl/strings/string_view.h"
+#include "quiche/common/simple_buffer_allocator.h"
+#include "quiche/quic/core/http/quic_spdy_stream.h"
+
+namespace Envoy {
+namespace Quic {
+
+HttpDatagramHandler::HttpDatagramHandler(quic::QuicSpdyStream* stream) : stream_(stream) {
+  ASSERT(stream != nullptr);
+  stream_->RegisterHttp3DatagramVisitor(this);
+}
+
+HttpDatagramHandler::~HttpDatagramHandler() {
+  if (stream_ != nullptr) {
+    stream_->UnregisterHttp3DatagramVisitor();
+  }
+}
+
+void HttpDatagramHandler::decodeCapsule(const quiche::Capsule& capsule) {
+  quiche::QuicheBuffer serialized_capsule = SerializeCapsule(capsule, &capsule_buffer_allocator_);
+  Buffer::InstancePtr buffer = std::make_unique<Buffer::OwnedImpl>();
+  buffer->add(serialized_capsule.AsStringView());
+  if (!stream_decoder_) {
+    IS_ENVOY_BUG("HTTP/3 Datagram received before a stream decoder is set.");
+  }
+  stream_decoder_->decodeData(*buffer, stream_->IsDoneReading());
+}
+
+void HttpDatagramHandler::OnHttp3Datagram(quic::QuicStreamId stream_id, absl::string_view payload) {
+  ASSERT(stream_id == stream_->id());
+  ENVOY_LOG(debug, "received a HTTP/3 Datagram with stream id={}", stream_id);
+  decodeCapsule(quiche::Capsule::Datagram(payload));
+}
+
+void HttpDatagramHandler::OnUnknownCapsule(quic::QuicStreamId stream_id,
+                                           const quiche::UnknownCapsule& capsule) {
+  ASSERT(stream_id == stream_->id());
+  ENVOY_LOG(debug, "received an Unknown Capsule with stream id={}", stream_id);
+  decodeCapsule(quiche::Capsule(capsule));
+}
+
+bool HttpDatagramHandler::OnCapsule(const quiche::Capsule& capsule) {
+  quiche::CapsuleType capsule_type = capsule.capsule_type();
+  if (capsule_type != quiche::CapsuleType::DATAGRAM) {
+    // Forward other types of Capsules without modifications.
+    stream_->WriteCapsule(capsule, fin_set_);
+    return true;
+  }
+  quic::MessageStatus status =
+      stream_->SendHttp3Datagram(capsule.datagram_capsule().http_datagram_payload);
+  // Returns false and thus resets the corresponding stream in the following statuses. Otherwise,
+  // drops the Datagram and move on without reporting a failure.
+  if (status == quic::MessageStatus::MESSAGE_STATUS_ENCRYPTION_NOT_ESTABLISHED ||
```

Comment:
Hm. it's a bit surprising to see a list of error codes here. I wonder if this belongs in a helper function of some sort. In particular, UNSUPPORTED and BLOCKED do not count as "errors" here. How come? maybe we should invert this logic and say `if (SUCCESS || UNSUPPORTED || BLOCKED) return true;` in particular, I'm apprehensive that we might add some other new fatal error in the future and not update this if.

Reply:
So this is based on @DavidSchinazi 's suggestion that datagrams blocked by the congestion control and size limit  should be simply dropped rather than being reported as errors. ( https://github.com/envoyproxy/envoy/pull/25925#discussion_r1136419184 ). Btw, the UNSUPPORTED status is consdiered as an error here already.

I think inverting the logic for the reason Ryan pointed out makes sense. Let me know if have any other thoughts @DavidSchinazi.